### PR TITLE
8344171: Clone and initialize Assertion Predicates in order instead of in reverse-order

### PR DESCRIPTION
**Pull request by chhagedorn**
This patch changes the order in which we clone and initialize Assertion Predicates from "reverse-order" to "in-order".

#### Current State: Mostly "reverse-order" for Assertion Predicates
We are currently cloning and initializing Assertion Predicates in reverse-order out of convenience and simplicity for most of the loop splitting optimizations - except for Loop Unswitching (see next section). This means that we do the following:
```
                                     old target loop entry
                                               |
         x                         Cloned Template Assertion
         |                                Predicate 2
Template Assertion                             |
    Predicate 1                       Initialized Assertion
         |               ==>               Predicate 2
Template Assertion                             |
    Predicate 2                    Cloned Template Assertion
         |                                Predicate 1
    source loop                                |
                                      Initialized Assertion
                                           Predicate 1
                                               |
                                          target loop
```
I don't think this is wrong but still kinda unexpected when trying to reason about a graph. But now with the recent refactorings, I think it's easy to change this to an in-order processing:
```
                                     old target loop entry
                                               |
         x                          Cloned Template Assertion
         |                                Predicate 1
Template Assertion                             |
    Predicate 1                       Initialized Assertion
         |               ==>               Predicate 1
Template Assertion                             |
    Predicate 2                     Cloned Template Assertion
         |                                Predicate 2
    source loop                                |
                                      Initialized Assertion
                                           Predicate 2
                                               |
                                          target loop
```
 This will also align all cloning/initializing of Assertion Predicates to the same order which was not the case before: Loop Unswitching already had an in-order cloning.

#### Why Does Loop Unswitching Use In-Order?
The main reason was that we can reuse `create_new_if_for_predicate()` which allowed us to keep the UCT on the false path. Now that [JDK-8342047](https://bugs.openjdk.org/browse/JDK-8342047) is in, we no longer use this method because we are only using halt nodes on the false path and no UCTs anymore.

We could have flipped the order in Loop Unswitching to a reverse-order now. However, this is only possible for Assertion Predicates. Parse Predicates, which are also cloned during Loop Unswitching, still must keep their relative order. To do the cloning of Parse Predicates with the same predicate visitor and simplify the reasoning about a graph, I propose to switch to an in-order cloning/initialization for all the Assertion Predicates.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344171](https://bugs.openjdk.org/browse/JDK-8344171): Clone and initialize Assertion Predicates in order instead of in reverse-order (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [dfcbf459](https://git.openjdk.org/jdk/pull/22275/files/dfcbf459696e43396d747412dd8561e1a418062c)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22275/head:pull/22275` \
`$ git checkout pull/22275`

Update a local copy of the PR: \
`$ git checkout pull/22275` \
`$ git pull https://git.openjdk.org/jdk.git pull/22275/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22275`

View PR using the GUI difftool: \
`$ git pr show -t 22275`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22275.diff">https://git.openjdk.org/jdk/pull/22275.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22275#issuecomment-2488500073)
</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -1753,20 +1753,11 @@ void PhaseIdealLoop::create_assertion_predicates_at_loop(CountedLoopNode* source
...
+  CreateAssertionPredicatesVisitor create_assertion_predicates_visitor(target_loop_head, this, _node_in_loop_body,
+                                                                       clone_template);
   Node* source_loop_entry = source_loop_head->skip_strip_mined()->in(LoopNode::EntryControl);
   PredicateIterator predicate_iterator(source_loop_entry);
   predicate_iterator.for_each(create_assertion_predicates_visitor);
-  if (create_assertion_predicates_visitor.has_created_predicates()) {
-    IfTrueNode* last_created_predicate_success_proj = create_assertion_predicates_visitor.last_created_success_proj();
-    _igvn.replace_input_of(target_outer_loop_head, LoopNode::EntryControl, last_created_predicate_success_proj);
-    set_idom(target_outer_loop_head, last_created_predicate_success_proj, dom_depth(target_outer_loop_head));
-  }
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:1769]: Workflow before was: Reverse-order cloning, so the very last clone needs to be connected to the loop head which was done here. With in-order, we can take care of the rewiring in the predicate visitor itself.

**Review by vnkozlov — APPROVED** [review]
Seems fine to me.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

**Review by eme64 — APPROVED** [review]
Nice refactoring, looks good!

**Comment by chhagedorn** [comment]
Accidentally pushed the next patch to the wrong branch.

**Review by eme64 — APPROVED** [review]

**Comment by chhagedorn** [comment]
Ran some testing again with latest master which looked good.

/integrate

**Pull request closed without merging**